File: ABC/249/c.cpp
/* 問題文
英小文字のみからなる N 個の文字列 S1,S2,…,SN が与えられます。
S1,S2,…,SN  から文字列を好きな個数選ぶことを考えます。
このとき、「選んだ文字列の中でちょうど K 個の文字列に登場する英小文字」の種類数としてあり得る最大値を求めてください。

制約
1≤N≤15
1≤K≤N
N,K は整数
Si は英小文字からなる空でない文字列である。
1≤i≤N を満たす整数 i に対し、Si  に同じ文字は 2 個以上含まれない。
i !=j ならばSi != Sj  である。*/

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n, k;
    cin >> n >> k;

    vector<string> s(n);
    for (string& str : s) cin >> str;

    int maxOccurrence = 0;

    // 各文字列は、選ぶ/選ばないの２通りあるので、N個の文字列からいくつか選ぶ方法は2^N通りある。
    // ビット全探索で全て調べ上げて解く。（Nが非常に小さいので可能）
    for (int bit = 0; bit < (1 << n); ++bit) {
        vector<int> charCount(26, 0);
        for (int idx = 0; idx < n; ++idx) {
            // idx番目の文字列を選ぶ
            if ((bit >> idx) & 1) {
                // 各文字の出現数を調べてインクリメント
                for (char c : s[idx]) charCount[c - 'a']++;
            }
        }
        // 目的の出現数を達成した文字の数を計算
        int currentOccurrence = count(charCount.begin(), charCount.end(), k);

        // 最大出現数を更新
        maxOccurrence = max(maxOccurrence, currentOccurrence);
    }
    cout << maxOccurrence << endl;

    return 0;
}

File: ABC/249/c.js
const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let input = [];
rl.on('line', (line) => {
    input.push(line);
}).on('close', () => {
    const [n, k] = input[0].split(' ').map(Number);
    const strings = input.slice(1, n + 1);

    let maxOccurrence = 0;
    for (let bit = 0; bit < (1 << n); ++bit) {
        const charCount = new Array(26).fill(0);
        for (let idx = 0; idx < n; ++idx) {
            if ((bit >> idx) & 1) {
                for (let char of strings[idx]) {
                    charCount[char.charCodeAt(0) - 'a'.charCodeAt(0)]++;
                }
            }
        }
        let currentOccurrence = charCount.filter(count => count === k).length;
        maxOccurrence = Math.max(maxOccurrence, currentOccurrence);
    }
    console.log(maxOccurrence);
});
